src/native/libs/System.Security.Cryptography.Native/memory_debug.c

```diff
@@ -0,0 +1,363 @@
+// there may be multiple mallocs in flight) and EnableMemoryTracking uses the
+// write side (to exclude all other accesses). The RW lock effectively protects
+// the g_trackingEnabledSince variable
+//
+
+// helper macro for getting the pointer to containing type from a pointer to a member
+#define container_of(ptr, type, member) ((type*)((char*)(ptr) - offsetof(type, member)))
+
+typedef struct link_t
+{
+    struct link_t* next;
+    struct link_t* prev;
+} link_t;
+
+typedef struct list_t
+{
+    link_t head;
+    pthread_mutex_t lock;
+} list_t;
+
+// Whether outstanding allocations should be tracked, UINT64_MAX if tracking is
+// disabled.
+static uint64_t g_trackingEnabledSince = UINT64_MAX;
+// Static lists of tracked allocations.
+static list_t* g_trackedMemory = NULL;
+// Number of partitions (distinct lists in g_trackedMemory) to track outstanding
+// allocations in.
+static uint32_t kPartitionCount = 32;
+// RW lock protecting g_trackingEnabledSince and g_trackedMemory.
+static pthread_rwlock_t g_trackedMemoryLock;
+
+// header for each tracked allocation
+typedef struct header_t
+{
+    // link for the circular list
+    struct link_t link;
+
+    // ordinal number of the allocation. Used to determine the target partition.
+    uint64_t index;
+
+    // size of the allocation (of the data field)
+    uint64_t size;
+
+    // filename from where the allocation was made
+    const char* file;
+
+    // line number in the file where the allocation was made
+    int32_t line;
+
+    // the start of actual allocated memory
+    __attribute__((aligned(8))) uint8_t data[];
+} header_t;
+
+static CRYPTO_RWLOCK* g_allocLock = NULL;
+static uint64_t g_allocatedMemory;
+static uint64_t g_allocationCount;
+
+static void list_link_init(link_t* link)
+{
+    link->next = link;
+    link->prev = link;
+}
+
+static void list_init(list_t* list)
+{
+    list_link_init(&list->head);
+
+    int res = pthread_mutex_init(&list->lock, NULL);
+    assert(res == 0);
+}
+
+static void list_insert_link(link_t* item, link_t* prev, link_t* next)
+{
+    next->prev = item;
+    item->next = next;
+    item->prev = prev;
+    prev->next = item;
+}
+
+static void list_unlink_item(link_t* item)
+{
+    link_t* prev = item->prev;
+    link_t* next = item->next;
+
+    prev->next = next;
+    next->prev = prev;
+
+    list_link_init(item);
+}
+
+static void init_memory_entry(header_t* entry, size_t size, const char* file, int32_t line, uint64_t index)
+{
+    entry->size = size;
+    entry->line = line;
+    entry->file = file;
+    list_link_init(&entry->link);
+    entry->index = index;
+}
+
+static uint64_t atomic_add64(uint64_t* value, uint64_t addend, CRYPTO_RWLOCK* lock)
+{
+    if (API_EXISTS(CRYPTO_atomic_add64))
+    {
+        uint64_t result;
+        CRYPTO_atomic_add64(value, addend, &result, lock);
+        return result;
+    }
+
+    // TODO: test other compilers, solve for 32-bit platforms.
```

Comment:
What's there to solve for 32-bit platforms?

I would expect `__atomic_fetch_add` to work fine everywhere. I think it would be fine to delete `CRYPTO_atomic_add64` plumbing and just use `__atomic_fetch_add` everywhere.

Reply:
I was not sure if 64bit atomics exist on 32bit platforms (or why else would CRYPTO_atomic_add(64) take a lock?)

Reply:
64bit atomics must effectively exist on all platforms that .NET runs on. We depend on support for 64bit atomics in many places.

> why else would CRYPTO_atomic_add(64) take a lock?

to support old platforms that we do not care about